File: backend/app/schemas/preview.py
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class RuntimeGraphNodeType(StrEnum):
    ORCHESTRATOR = "orchestrator"
    TOOL = "tool"
    WORKFLOW = "workflow"
    SUB_AGENT = "sub_agent"
    KNOWLEDGE_BASE = "knowledge_base"


class RuntimeGraphEdgeKind(StrEnum):
    TOOL = "tool"
    WORKFLOW = "workflow"
    SUB_AGENT = "sub_agent"
    KNOWLEDGE_BASE = "knowledge_base"


class RuntimeGraphOrchestrator(BaseModel):
    id: str
    name: str
    kind: str = "orchestrator"
    status: str | None = None
    description: str | None = None


class RuntimeGraphNode(BaseModel):
    id: str
    type: RuntimeGraphNodeType
    label: str
    description: str | None = None


class RuntimeGraphEdge(BaseModel):
    id: str
    source: str
    target: str
    kind: RuntimeGraphEdgeKind


class RuntimeGraphResponse(BaseModel):
    agent_id: str
    organization_id: str
    orchestrator: RuntimeGraphOrchestrator
    nodes: list[RuntimeGraphNode] = Field(default_factory=list)
    edges: list[RuntimeGraphEdge] = Field(default_factory=list)


class PreviewTraceEvent(BaseModel):
    type: str
    at: str
    data: dict[str, Any] = Field(default_factory=dict)


class PreviewTraceTurn(BaseModel):
    turn_id: str
    started_at: str
    events: list[PreviewTraceEvent] = Field(default_factory=list)
    routing_decision: dict[str, Any] | None = None


class PreviewTraceResponse(BaseModel):
    agent_id: str
    sender_id: str
    source: str | None = None
    turns: list[PreviewTraceTurn] = Field(default_factory=list)

File: frontend/src/types/preview.ts
export type RuntimeGraphNodeType =
  | "orchestrator"
  | "tool"
  | "workflow"
  | "sub_agent"
  | "knowledge_base";

export type RuntimeGraphEdgeKind = "tool" | "workflow" | "sub_agent" | "knowledge_base";

export type RuntimeGraphOrchestrator = {
  id: string;
  name: string;
  kind: string;
  status?: string | null;
  description?: string | null;
};

export type RuntimeGraphNode = {
  id: string;
  type: RuntimeGraphNodeType;
  label: string;
  description: string | null;
};

export type RuntimeGraphEdge = {
  id: string;
  source: string;
  target: string;
  kind: RuntimeGraphEdgeKind;
};

export type RuntimeGraphResponse = {
  agent_id: string;
  organization_id: string;
  orchestrator: RuntimeGraphOrchestrator;
  nodes: RuntimeGraphNode[];
  edges: RuntimeGraphEdge[];
};

export type PreviewTraceEvent = {
  type: string;
  at: string;
  data: Record<string, unknown>;
};

export type PreviewTraceTurn = {
  turn_id: string;
  started_at: string;
  events: PreviewTraceEvent[];
  routing_decision: Record<string, unknown> | null;
};

export type PreviewTraceResponse = {
  agent_id: string;
  sender_id: string;
  source: string | null;
  turns: PreviewTraceTurn[];
};

export type ChatButton = {
  title: string;
  payload: string;
};

export type ChatMessage = {
  recipient_id: string;
  text: string | null;
  buttons: ChatButton[] | null;
};

export type PreviewChatRequest = {
  sender_id: string;
  message: string;
  metadata?: Record<string, unknown>;
};

export type PreviewChatResponse = {
  messages: ChatMessage[];
};

export type PositionedRuntimeGraphNode = RuntimeGraphNode & {
  position: { x: number; y: number };
};
